frontend/docsettings.lua

```diff
@@ -527,4 +615,71 @@ function DocSettings:flushCustomMetadata(doc_path)
     end
 end
 
+-- hash-based SDR storage
+
+local function getSdrsInDir(path)
+    -- Get all the metadata.filetype.lua files under directory path.
+    -- Derived from readerdictionary.getIfosInDir()
+    local sdrs = {}
+    local ok, iter, dir_obj = pcall(lfs.dir, path)
+    if ok then
+        for name in iter, dir_obj do
+            if name ~= "." and name ~= ".." then
+                local fullpath = path.."/"..name
+                local attributes = lfs.attributes(fullpath)
+                if attributes ~= nil then
+                    if attributes.mode == "directory" then
+                        local dirifos = getSdrsInDir(fullpath) -- recurse
+                        for _, ifo in pairs(dirifos) do
+                            table.insert(sdrs, ifo)
+                        end
+                    elseif fullpath:match("metadata%..+%.lua$") then
+                        table.insert(sdrs, fullpath)
+                    end
+                end
+            end
+        end
+    end
+    return sdrs
+end
+
+
+function DocSettings.getHashDirSdrInfos()
+    local sdrs = getSdrsInDir(DOCSETTINGS_HASH_DIR)
+    local title_author_strs = {}
+    for _, sdr in ipairs(sdrs) do
+        -- Ignore empty files
+        if lfs.attributes(sdr, "size") > 0 then
+            local ok, stored
+            ok, stored = pcall(dofile, sdr)
+            -- Ignore empty tables
+            if ok and next(stored) ~= nil then
+                local info_str = stored.doc_props.title
+                if not info_str then info_str = stored.stats.title end
```

Comment:
`stats` takes metadata from `doc_props`, so this would not help.

Reply:
removed this and for authors

Reply:
@hius07 I have some documents that have `stats.title` but not `doc_props.title` so I'm inclined to put this back in

Reply:
It's just a filename without extension, you can get it without the Stats plugin.
https://github.com/koreader/koreader/blob/2ff60ce04f6bca15e7c4974588dd10df5cb65575/plugins/statistics.koplugin/main.lua#L178
https://github.com/koreader/koreader/blob/2ff60ce04f6bca15e7c4974588dd10df5cb65575/frontend/apps/filemanager/filemanagerbookinfo.lua#L195

Reply:
BTW, the list of "hash" books doesn't take into account custom metadata.
But in the file browser a user will see titles/authors customized.

Reply:
Thanks, if it's just from the filename we're good then.

if there's a specific additional field to check for each of title and author for custom metadata I'm happy to check for it first. Really this is for a very niche use case so it's just meant to give some basic info to help users identify if they care to move the metadata themselves vs. delete but it's easy to add in more things.

Reply:
figured it out and added these in